### PR TITLE
session-loader: Don't loop forever when we don't find a file in any multi component

### PR DESCRIPTION
**Pull request by wz1000**
We add a check for if the current file is a target we know about, and emit a
diagnostic if that is the case, refusing to load the file in.

This doesn't change the implicit adding of the current file as a target for a
single component case, as we need the old behaviour to support bare GHC/Direct
cradles where not all targets may be listed.

TODO: add test case

### CODE REVIEW

**Comment by michaelpj** [comment]
Is this related to https://github.com/haskell/haskell-language-server/issues/4046 ?

**Comment by wz1000** [comment]
> Is this related to #4046 ?

I don't think so

**Review by michaelpj — APPROVED** [review]
Seems reasonable, test case would be great.

File: ghcide/session-loader/Development/IDE/Session.hs [hunk 1]
```diff
@@ -585,9 +585,21 @@ loadSessionWithOptions recorder SessionLoadingOptions{..} dir = do
...
+          let (all_targets, this_flags_map, this_options)
+                = case HM.lookup _cfp flags_map' of
+                    Just this -> (all_targets', flags_map', this)
+                    Nothing -> (this_target_details : all_targets', HM.insert _cfp this_flags flags_map', this_flags)
+                  where all_targets' = concat all_target_details
+                        flags_map' = HM.fromList (concatMap toFlagsMap all_targets')
+                        this_target_details = TargetDetails (TargetFile _cfp) this_error_env this_dep_info [_cfp]
+                        this_flags = (this_error_env, this_dep_info)
+                        this_error_env = ([this_error], Nothing)
+                        this_error = ideErrorWithSource (Just "crade") (Just DiagnosticSeverity_Error) _cfp
```
Comment (fendor) [inline ghcide/session-loader/Development/IDE/Session.hs:598]:
```suggestion
                        this_error = ideErrorWithSource (Just "cradle") (Just DiagnosticSeverity_Error) _cfp
```

**Review by fendor — APPROVED** [review]
LGTM

**Pull request merged**